### PR TITLE
Support to apply dml msgstream from pool

### LINKED ISSUE #6954: Response time of Flush, Search and Query increased 10+ times if there are hundreds of collections in a Milvus cluster
#### Steps/Code to reproduce:
1. deploy a Milvus with cluster mode
2. run  pytest testcases/test_e2e.py --host x.x.x.x in /milvus/tests20/python_client
3. check the test log: every request responses in ~3s:
[2021-08-02 17:52:24,151 - DEBUG - ci_test]: assert create: 0:00:01.131757 (test_e2e.py:23)
[2021-08-02 17:52:24,885 - DEBUG - ci_test]: assert insert: 0:00:00.695129 (test_e2e.py:31)
[2021-08-02 17:52:28,057 - DEBUG - ci_test]: assert flush: 0:00:03.172252 (test_e2e.py:38)
[2021-08-02 17:52:28,761 - DEBUG - ci_test]: assert search: 0:00:00.126437 (test_e2e.py:48)
[2021-08-02 17:52:35,412 - DEBUG - ci_test]: assert index: 0:00:02.172020 (test_e2e.py:61)
[2021-08-02 17:52:35,446 - DEBUG - ci_test]: assert query: 0:00:00.020285 (test_e2e.py:69)

4. create about 400 empty collections 
5. repeat step#2
6. check the new test logs
[2021-08-02 18:52:33,450 - DEBUG - ci_test]: assert create: 0:00:02.133666 (test_e2e.py:23)
[2021-08-02 18:52:34,066 - DEBUG - ci_test]: assert insert: 0:00:00.568658 (test_e2e.py:31)
[2021-08-02 18:53:14,578 - DEBUG - ci_test]: **assert flush: 0:00:40.511621** (test_e2e.py:38)
[2021-08-02 18:53:51,154 - DEBUG - ci_test]: **assert search: 0:00:35.778467** (test_e2e.py:49)
[2021-08-02 18:54:38,310 - DEBUG - ci_test]: assert index: 0:00:02.174828 (test_e2e.py:63)
[2021-08-02 18:55:10,895 - DEBUG - ci_test]: **assert query: 0:00:32.565820** (test_e2e.py:71)

#### Expected result:
response time shall be stable in about 3 seconds

#### Actual results:
response time of flush, search and query increased 10+ times

#### Environment:
- Milvus version(e.g. v2.0.0-RC2 or 8b23a93): master-20210729-7307332
- Deployment mode(standalone or cluster): clust
- SDK version(e.g. pymilvus v2.0.0rc2): 
pymilvus                           2.0.0rc3.dev15
pymilvus-orm                       2.0.0rc3.dev13
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:

### PR DESCRIPTION
**Pull request by cydrain**
Signed-off-by: yudong.cai <yudong.cai@zilliz.com>

Resolves: #6954

### CODE REVIEW

**Review by XuanYang-cn — COMMENTED** [review]
/lgtm

File: internal/rootcoord/dml_channels.go [hunk 1]
```diff
@@ -15,25 +15,61 @@ import (
...
+	refcnt     map[string]int64
+	namePrefix string
+	idx        int
+	idxLock    sync.Mutex
+	capacity   int
+	pool       map[string]msgstream.MsgStream
 }
 
 func newDMLChannels(c *Core) *dmlChannels {
+	const totalDmlMsgStreamNum = 64
```
Comment (xiaofan-luan) [inline internal/rootcoord/dml_channels.go:38]: should not be a magic number here. 
We will need a configuration some where.

File: internal/rootcoord/dml_channels.go [hunk 2]
```diff
@@ -15,25 +15,61 @@ import (
...
+	"github.com/milvus-io/milvus/internal/util/funcutil"
 )
 
 type dmlChannels struct {
-	core *Core
-	lock sync.RWMutex
-	dml  map[string]msgstream.MsgStream
+	core       *Core
+	lock       sync.RWMutex
+	dml        map[string]msgstream.MsgStream
```
Comment (xiaofan-luan) [inline internal/rootcoord/dml_channels.go:28]: remove dml map

File: internal/rootcoord/dml_channels.go [hunk 3]
```diff
@@ -15,25 +15,61 @@ import (
...
+		refcnt:     make(map[string]int64),
+		namePrefix: namePrefix,
+		idx:        0,
+		capacity:   totalDmlMsgStreamNum,
+		pool:       pool,
 	}
 }
 
+func (d *dmlChannels) GetDmlMsgStreamName() string {
+	d.idxLock.Lock()
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/rootcoord/dml_channels.go:66]: use atomic.Int64 instead

File: internal/querynode/flow_graph_query_node.go [hunk 4]
```diff
@@ -125,6 +125,7 @@ func (q *queryNodeFlowGraph) consumerFlowGraph(channel Channel, subName ConsumeS
 	log.Debug("query node flow graph consumes from pChannel",
 		zap.Any("collectionID", q.collectionID),
 		zap.Any("channel", channel),
+		zap.Any("subName", subName),
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/querynode/flow_graph_query_node.go:128]: already comment this log

File: internal/datanode/flow_graph_dd_node.go [hunk 5]
```diff
@@ -92,15 +92,22 @@ func (ddn *ddNode) Operate(in []flowgraph.Msg) []flowgraph.Msg {
...
 					zap.Uint64("Message endts", msg.EndTs()),
 					zap.Uint64("FilterThreshold", FilterThreshold),
 				)
-				if ddn.filterFlushedSegmentInsertMessages(msg.(*msgstream.InsertMsg)) {
+				if ddn.filterFlushedSegmentInsertMessages(imsg) {
 					continue
 				}
 			}
+			if imsg.CollectionID != ddn.collectionID {
+				log.Debug("filter invalid InsertMsg, collection mis-match",
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/datanode/flow_graph_dd_node.go:106]: already comment this log

File: internal/rootcoord/dml_channels.go (hunk 1 above)
Comment (cydrain) [reply internal/rootcoord/dml_channels.go:38]: add config instead

File: internal/rootcoord/dml_channels.go (hunk 2 above)
Comment (cydrain) [reply internal/rootcoord/dml_channels.go:28]: done

File: internal/rootcoord/timeticksync.go [hunk 6]
```diff
@@ -243,16 +238,16 @@ func (t *timetickSync) StartWatch() {
...
 			}
 		}
 	}
 }
 
-// SendChannelTimeTick send each channel's min timetick to msg stream
-func (t *timetickSync) SendChannelTimeTick(chanName string, ts typeutil.Timestamp) error {
+// SendTimeTickToChannel send each channel's min timetick to msg stream
+func (t *timetickSync) SendTimeTickToChannel(chanNames []string, ts typeutil.Timestamp) error {
 	msgPack := msgstream.MsgPack{}
```
[the remark this answers is not in the document]
Comment (cydrain) [reply internal/rootcoord/timeticksync.go:267]: done

**Pull request merged**